src/solver/td3.ml

```diff
@@ -395,90 +438,62 @@ module Base =
-          | "sides-local" when not (S.Dom.leq b a) -> (
-              match x with
-              | None -> widen a b
-              | Some x when VS.mem x old_sides -> widen a b
-              | _ -> S.Dom.join a b
-            )
-          | _ when HM.mem wpoint y  -> widen a b
-          | _ -> S.Dom.join a b
+        let vetoed_widen = WPS.veto_widen wps_data called old_sides x y in
+        let op a b = (* If y still has widening gas, widening will not be performed. *)
+          if vetoed_widen || not (should_widen y) then S.Dom.join a b else widen a b
         in
         let old = HM.find rho y in
         let tmp = op old d in
         if tracing then trace "sol2" "stable add %a" S.Var.pretty_trace y;
         HM.replace stable y ();
         if not (S.Dom.leq tmp old) then (
-          if tracing && not (S.Dom.is_bot old) then trace "solside" "side to %a (wpx: %b) from %a: %a -> %a" S.Var.pretty_trace y (HM.mem wpoint y) (Pretty.docOpt (S.Var.pretty_trace ())) x S.Dom.pretty old S.Dom.pretty tmp;
-          if tracing && not (S.Dom.is_bot old) then trace "solchange" "side to %a (wpx: %b) from %a: %a" S.Var.pretty_trace y (HM.mem wpoint y) (Pretty.docOpt (S.Var.pretty_trace ())) x S.Dom.pretty_diff (tmp, old);
-          let sided = match x with
-            | Some x ->
-              let sided = VS.mem x old_sides in
-              if not sided then add_sides y x;
-              sided
-            | None -> false
-          in
+          if tracing && not (S.Dom.is_bot old) then trace "solside" "side to %a (wpx: %s) from %a: %a -> %a" S.Var.pretty_trace y (format_wpoint y) (Pretty.docOpt (S.Var.pretty_trace ())) x S.Dom.pretty old S.Dom.pretty tmp;
+          if tracing && not (S.Dom.is_bot old) then trace "solchange" "side to %a (wpx: %s) from %a: %a" S.Var.pretty_trace y (format_wpoint y) (Pretty.docOpt (S.Var.pretty_trace ())) x S.Dom.pretty_diff (tmp, old);
+
+          (match x with
+           | Some x ->
+             if not (VS.mem x old_sides) then add_sides y x;
+           | None -> ());
+
           (* HM.replace rho y ((if HM.mem wpoint y then S.Dom.widen old else identity) (S.Dom.join old d)); *)
           HM.replace rho y tmp;
-          if side_widen <> "cycle" then destabilize y;
+          let destabilized_vs: bool option = if WPS.record_destabilized_vs then (
+              destabilize y;
+              None
+            ) else
+              Some (destabilize_vs y) in
+
           (* make y a widening point if ... This will only matter for the next side _ y.  *)
-          let wpoint_if e =
-            if e then (
-              if tracing then trace "sol2" "side adding wpoint %a from %a" S.Var.pretty_trace y (Pretty.docOpt (S.Var.pretty_trace ())) x;
-              HM.replace wpoint y ()
-            )
-          in
-          match side_widen with
-          | "always" -> (* Any side-effect after the first one will be widened which will unnecessarily lose precision. *)
```

Comment:
These comments should not be removed but moved to the corresponding extracted modules.